File: 13. Longest Substring with K Unique Characters.cpp
// Topic: Longest Substring with K Unique Characters
#include <bits/stdc++.h>
using namespace std;

int main()
{
    string s = "araaci";
    int k = 2;
    unordered_map<char, int> freq;
    int left = 0, maxlen = 0;
    for (int right = 0; right < s.size(); ++right)
    {
        freq[s[right]]++;
        while (freq.size() > k)
        {
            if (--freq[s[left]] == 0)
                freq.erase(s[left]);
            left++;
        }
        maxlen = max(maxlen, right - left + 1);
    }
    cout << "Longest: " << maxlen << '\n';
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(k)

File: 13. Longest Substring with K Unique Characters.py
# Topic: Longest Substring with K Unique Characters

def longest_substring_k_unique(s, k):
    freq = {}
    left = 0
    maxlen = 0
    for right, ch in enumerate(s):
        freq[ch] = freq.get(ch, 0) + 1
        while len(freq) > k:
            freq[s[left]] -= 1
            if freq[s[left]] == 0:
                del freq[s[left]]
            left += 1
        maxlen = max(maxlen, right-left+1)
    return maxlen

if __name__ == '__main__':
    s = "araaci"
    k = 2
    print("Longest:", longest_substring_k_unique(s, k))

# Time Complexity: O(n)
# Space Complexity: O(k)
